### ISSUE TITLE
build error with grpc-swift dependency

### ISSUE THREAD

**Issue by oliverepper** [grpc/grpc-swift#1540]
I am trying to build a project of mine that depends on grpc-swift and get a lot of these errors:

error: value of type 'ChannelPipeline.SynchronousOperations' has no member 'nioSSL_tlsVersion'
`return try self.pipeline.syncOperations.nioSSL_tlsVersion().map(GRPCTLSVersion.init)`

This happens while depending on version 1.13.1. At the moment I have no idea what I could change.

**Comment by dnadoba** [comment]
You need a newer version of `swift-nio`. We forgot to update the minimum required swift version in this PR: https://github.com/grpc/grpc-swift/pull/1509

For now you can just update to the latest swift-nio version to resolve the issue manually.

However, we should update our Package.swift too.

**Comment by oliverepper** [comment]
Thanks for your help! Shall I make a PR for this? Or is there anything special about the release process? If it's not fixed upstream I need to fix this in my branch, anyways. It needed updating of swift-nio-ssl, btw.

**Comment by Lukasa** [comment]
A PR for this would be very welcome.

**Comment by oliverepper** [comment]
For now the PR would only be the update of the version of the swift-nio-ssl dependency. I could go through all of the dependencies and try updating them but right now I have trouble at runtime. I just want to make sure that the trouble is not related. This is the error I get:

`[...]/Sources/NIOExtras/QuiescingHelper.swift:222: Fatal error: leaking promise created at (file: "/MicroSwitch/.build/checkouts/swift-nio-extras/Sources/NIOExtras/QuiescingHelper.swift", line: 222)`

I'll try updating the other nio dependencies as well.

**Comment by Lukasa** [comment]
We should only need to uplift the swift-nio-ssl dependency.

The issue you're seeing there is likely a follow-on from hitting an error condition: that runtime issue is unlikely to be fixed by lifting a NIO version. Did you see that issue after lifting the swift-nio-ssl dependency?

**Comment by oliverepper** [comment]
Oh. The error happens on Linux, only. MacOS is fine.

**Comment by Lukasa** [comment]
This crash is likely masking an error being thrown, which is causing the promise to leak. Can you produce a complete stack at the point of the crash?

**Comment by oliverepper** [comment]
I think I got it. Let me double check and if correct I will prepare the small PR for the new nio-ssl dependency.

**Comment by oliverepper** [comment]
Anyways here is the trace:

```pre
2023-01-03T09:22:26+0000 info mswitch : Using: MultiThreadedEventLoopGroup { threadPattern = NIO-ELT-0-#* }
/MicroSwitch/.build/checkouts/swift-nio-extras/Sources/NIOExtras/QuiescingHelper.swift:222: Fatal error: leaking promise created at (file: "/MicroSwitch/.build/checkouts/swift-nio-extras/Sources/NIOExtras/QuiescingHelper.swift", line: 222)
Current stack trace:
0    libswiftCore.so                    0x0000ffff9f79b354 _swift_stdlib_reportFatalErrorInFile + 128
1    libswiftCore.so                    0x0000ffff9f49aa1c <unavailable> + 1419804
2    libswiftCore.so                    0x0000ffff9f4997a4 _assertionFailure(_:_:file:line:flags:) + 304
3    mswitch                            0x0000aaaabd1648c0 <unavailable> + 9062592
4    mswitch                            0x0000aaaabd1aa0b4 <unavailable> + 9347252
5    mswitch                            0x0000aaaabd1aa014 <unavailable> + 9347092
6    mswitch                            0x0000aaaabd1642d8 <unavailable> + 9061080
7    mswitch                            0x0000aaaabd164b0c <unavailable> + 9063180
8    libswiftCore.so                    0x0000ffff9f72536c <unavailable> + 4084588
9    libswiftCore.so                    0x0000ffff9f725fb0 <unavailable> + 4087728
10   mswitch                            0x0000aaaabd1cae2c <unavailable> + 9481772
11   mswitch                            0x0000aaaabd1cae04 <unavailable> + 9481732
12   mswitch                            0x0000aaaabd1cae58 <unavailable> + 9481816
13   libswiftCore.so                    0x0000ffff9f72536c <unavailable> + 4084588
14   libswiftCore.so                    0x0000ffff9f725fb0 <unavailable> + 4087728
15   mswitch                            0x0000aaaabd03a4a0 <unavailable> + 7840928
16   libswiftCore.so                    0x0000ffff9f72536c <unavailable> + 4084588
17   libswiftCore.so                    0x0000ffff9f725fb0 <unavailable> + 4087728
18   mswitch                            0x0000aaaabd171594 <unavailable> + 9115028
19   libswiftCore.so                    0x0000ffff9f72536c <unavailable> + 4084588
20   libswiftCore.so                    0x0000ffff9f725fb0 <unavailable> + 4087728
21   mswitch                            0x0000aaaabd171458 <unavailable> + 9114712
22   libswiftCore.so                    0x0000ffff9f72536c <unavailable> + 4084588
23   libswiftCore.so                    0x0000ffff9f725fb0 <unavailable> + 4087728
24   mswitch                            0x0000aaaabd3e8aac <unavailable> + 11700908
25   libswiftCore.so                    0x0000ffff9f72536c <unavailable> + 4084588
26   libswiftCore.so                    0x0000ffff9f725fb0 <unavailable> + 4087728
27   libswiftCore.so                    0x0000ffff9f71aa70 swift_arrayDestroy + 68
28   libswiftCore.so                    0x0000ffff9f4d5948 ContiguousArray.replaceSubrange<A>(_:with:) + 1552
29   mswitch                            0x0000aaaabd3e3108 <unavailable> + 11677960
30   mswitch                            0x0000aaaabd3b203c <unavailable> + 11477052
31   mswitch                            0x0000aaaabd3b259c <unavailable> + 11478428
32   mswitch                            0x0000aaaabd419528 <unavailable> + 11900200
33   mswitch                            0x0000aaaabd41bdb4 <unavailable> + 11910580
34   mswitch                            0x0000aaaabd41bec4 <unavailable> + 11910852
35   libc.so.6                          0x0000ffff9ef1d5c8 <unavailable> + 513480
36   libc.so.6                          0x0000ffff9ef85d1c <unavailable> + 941340
Trace/breakpoint trap
```

**Comment by oliverepper** [comment]
How can I get more information out of this stack trace? I build with `swift build -c debug` and ran the resulting binary in gdb. I still don't see symbol names of the mswitch binary.

**Comment by oliverepper** [comment]
I see this in gdb:

`$s14ArgumentParser15ParsableCommandPAAE4mainyyFZ (self=...) at /MicroSwitch/.build/checkouts/swift-argument-parser/Sources/ArgumentParser/Parsable Types/ParsableCommand.swift:104`

followed by this:

`Thread 3 "NIO-ELT-0-#0" received signal SIGTRAP, Trace/breakpoint trap.
[Switching to Thread 0xffff97bdf060 (LWP 10855)]
0x0000ffff9aad98dc in $ss17_assertionFailure__4file4line5flagss5NeverOs12StaticStringV_SSAHSus6UInt32VtF () from /usr/lib/swift/linux/libswiftCore.so`

**Comment by Lukasa** [comment]
If you've got gdb attached, can you use lldb and type you type `bt` when you hit the SIGTRAP?

**Comment by oliverepper** [comment]
```pre
#0  0x0000ffff9fbf98dc in $ss17_assertionFailure__4file4line5flagss5NeverOs12StaticStringV_SSAHSus6UInt32VtF () from /usr/lib/swift/linux/libswiftCore.so
#1  0x0000aaaaafe75690 in $s7NIOCore15EventLoopFutureCfdyyXEfU_ () at /MicroSwitch/.build/checkouts/swift-nio/Sources/NIOCore/EventLoopFuture.swift:420
#2  0x0000aaaaafebae84 in $s7NIOCore9debugOnlyyyyyXEFSbyXEfu_SbyXEfU_ () at /MicroSwitch/.build/checkouts/swift-nio/Sources/NIOCore/Utilities.swift:45
#3  0x0000aaaaafebade4 in $s7NIOCore9debugOnlyyyyyXEF (body=...) at /MicroSwitch/.build/checkouts/swift-nio/Sources/NIOCore/Utilities.swift:45
#4  0x0000aaaaafe750a8 in $s7NIOCore15EventLoopFutureCfd (self=...) at /MicroSwitch/.build/checkouts/swift-nio/Sources/NIOCore/EventLoopFuture.swift:417
#5  0x0000aaaaafe758dc in $s7NIOCore15EventLoopFutureCfD (self=...) at /MicroSwitch/.build/checkouts/swift-nio/Sources/NIOCore/EventLoopFuture.swift:416
#6  0x0000ffff9fe8536c in _swift_release_dealloc () from /usr/lib/swift/linux/libswiftCore.so
#7  0x0000ffff9fe85fb0 in bool swift::RefCounts<swift::RefCountBitsT<(swift::RefCountInlinedness)1> >::doDecrementSlow<(swift::PerformDeinit)1>(swift::RefCountBitsT<(swift::RefCountInlinedness)1>, unsigned int) () from /usr/lib/swift/linux/libswiftCore.so
#8  0x0000aaaaafedbbfc in $s7NIOCore16EventLoopPromiseVy9NIOExtras16ChannelCollector33_15131B0A1C2B96335CB419BAF83D55DELLCGWOh ()
    at /MicroSwitch/.build/checkouts/swift-nio-extras/Sources/NIOExtras/QuiescingHelper.swift:214
#9  0x0000aaaaafedbbd4 in $s9NIOExtras21ServerQuiescingHelperCfd (self=...) at /MicroSwitch/.build/checkouts/swift-nio-extras/Sources/NIOExtras/QuiescingHelper.swift:214
#10 0x0000aaaaafedbc28 in $s9NIOExtras21ServerQuiescingHelperCfD (self=...) at /MicroSwitch/.build/checkouts/swift-nio-extras/Sources/NIOExtras/QuiescingHelper.swift:214
#11 0x0000ffff9fe8536c in _swift_release_dealloc () from /usr/lib/swift/linux/libswiftCore.so
#12 0x0000ffff9fe85fb0 in bool swift::RefCounts<swift::RefCountBitsT<(swift::RefCountInlinedness)1> >::doDecrementSlow<(swift::PerformDeinit)1>(swift::RefCountBitsT<(swift::RefCountInlinedness)1>, unsigned int) () from /usr/lib/swift/linux/libswiftCore.so
#13 0x0000aaaaafd4b2a0 in ?? () at /MicroSwitch/.build/checkouts/grpc-swift/Sources/GRPC/Server.swift:199
#14 0x0000ffff9fe8536c in _swift_release_dealloc () from /usr/lib/swift/linux/libswiftCore.so
#15 0x0000ffff9fe85fb0 in bool swift::RefCounts<swift::RefCountBitsT<(swift::RefCountInlinedness)1> >::doDecrementSlow<(swift::PerformDeinit)1>(swift::RefCountBitsT<(swift::RefCountInlinedness)1>, unsigned int) () from /usr/lib/swift/linux/libswiftCore.so
#16 0x0000aaaaafe82364 in ?? () at /MicroSwitch/.build/checkouts/swift-nio/Sources/NIOCore/EventLoopFuture.swift:1570
#17 0x0000ffff9fe8536c in _swift_release_dealloc () from /usr/lib/swift/linux/libswiftCore.so
#18 0x0000ffff9fe85fb0 in bool swift::RefCounts<swift::RefCountBitsT<(swift::RefCountInlinedness)1> >::doDecrementSlow<(swift::PerformDeinit)1>(swift::RefCountBitsT<(swift::RefCountInlinedness)1>, unsigned int) () from /usr/lib/swift/linux/libswiftCore.so
#19 0x0000aaaaafe82228 in ?? () at /MicroSwitch/.build/checkouts/swift-nio/Sources/NIOCore/EventLoopFuture.swift:1570
#20 0x0000ffff9fe8536c in _swift_release_dealloc () from /usr/lib/swift/linux/libswiftCore.so
#21 0x0000ffff9fe85fb0 in bool swift::RefCounts<swift::RefCountBitsT<(swift::RefCountInlinedness)1> >::doDecrementSlow<(swift::PerformDeinit)1>(swift::RefCountBitsT<(swift::RefCountInlinedness)1>, unsigned int) () from /usr/lib/swift/linux/libswiftCore.so
#22 0x0000aaaab00f987c in ?? () at /MicroSwitch/.build/checkouts/swift-nio/Sources/NIOPosix/SelectableEventLoop.swift:418
#23 0x0000ffff9fe8536c in _swift_release_dealloc () from /usr/lib/swift/linux/libswiftCore.so
#24 0x0000ffff9fe85fb0 in bool swift::RefCounts<swift::RefCountBitsT<(swift::RefCountInlinedness)1> >::doDecrementSlow<(swift::PerformDeinit)1>(swift::RefCountBitsT<(swift::RefCountInlinedness)1>, unsigned int) () from /usr/lib/swift/linux/libswiftCore.so
#25 0x0000ffff9fe7aab4 in swift_arrayDestroy () from /usr/lib/swift/linux/libswiftCore.so
#26 0x0000ffff9fc35f58 in $ss15ContiguousArrayV15replaceSubrange_4withySnySiG_qd__nt7ElementQyd__RszSlRd__lF () from /usr/lib/swift/linux/libswiftCore.so
#27 0x0000aaaab00f3ed8 in $s8NIOPosix19SelectableEventLoopC3runyyKF (self=...) at /MicroSwitch/.build/checkouts/swift-nio/Sources/NIOPosix/SelectableEventLoop.swift:513
#28 0x0000aaaab00c2e0c in $s8NIOPosix27MultiThreadedEventLoopGroupC06runTheE033_C2B1528F4FBA68A3DBFA89DBAEBE9D4DLL6thread03candE22BeShutdownIndividually15selectorFactory11initializer_yAA9NIOThreadC_SbAA8SelectorCyAA15NIORegistrationVGyKcyAKcyAA010SelectabledE0CctFZ (thread=..., canEventLoopBeShutdownIndividually=..., selectorFactory=..., initializer=..., 
    callback=..., self=...) at /MicroSwitch/.build/checkouts/swift-nio/Sources/NIOPosix/MultiThreadedEventLoopGroup.swift:87
#29 0x0000aaaab00c336c in $s8NIOPosix27MultiThreadedEventLoopGroupC014setupThreadAnddE033_C2B1528F4FBA68A3DBFA89DBAEBE9D4DLL4name15selectorFactory11initializerAA010SelectabledE0CSS_AA8SelectorCyAA15NIORegistrationVGyKcyAA9NIOThreadCctFZyAQcfU_ () at /MicroSwitch/.build/checkouts/swift-nio/Sources/NIOPosix/MultiThreadedEventLoopGroup.swift:107
#30 0x0000aaaab012a2f8 in $s8NIOPosix9NIOThreadCIegg_ACytIegnr_TR () at /MicroSwitch/.build/checkouts/swift-nio/Sources/NIOPosix/Thread.swift:97
#31 0x0000aaaab012cb84 in $s8NIOPosix14ThreadOpsPosixO3run6handle4args06detachB0ySuSgz_AA3BoxCyyAA9NIOThreadCc4body_SSSg4nametGSbtFZSvSgAQcfU_ ()
    at /MicroSwitch/.build/checkouts/swift-nio/Sources/NIOPosix/ThreadPosix.swift:105
#32 0x0000aaaab012cc94 in $s8NIOPosix14ThreadOpsPosixO3run6handle4args06detachB0ySuSgz_AA3BoxCyyAA9NIOThreadCc4body_SSSg4nametGSbtFZSvSgAQcfU_To ()
    at /MicroSwitch/.build/checkouts/swift-nio/Sources/NIOPosix/ThreadPosix.swift:102
#33 0x0000ffff9f67d5c8 in start_thread (arg=0x0) at ./nptl/pthread_create.c:442
#34 0x0000ffff9f6e5d1c in thread_start () at ../sysdeps/unix/sysv/linux/aarch64/clone.S:79
```

**Comment by Lukasa** [comment]
Demangled:

```
#0  0x0000ffff9fbf98dc in Swift._assertionFailure(_: Swift.StaticString, _: Swift.String, file: Swift.StaticString, line: Swift.UInt, flags: Swift.UInt32) -> Swift.Never () from /usr/lib/swift/linux/libswiftCore.so
#1  0x0000aaaaafe75690 in closure #1 () -> () in NIOCore.EventLoopFuture.deinit () at /MicroSwitch/.build/checkouts/swift-nio/Sources/NIOCore/EventLoopFuture.swift:420
#2  0x0000aaaaafebae84 in closure #1 () -> Swift.Bool in implicit closure #1 () -> Swift.Bool in NIOCore.debugOnly(() -> ()) -> () () at /MicroSwitch/.build/checkouts/swift-nio/Sources/NIOCore/Utilities.swift:45
#3  0x0000aaaaafebade4 in NIOCore.debugOnly(() -> ()) -> () (body=...) at /MicroSwitch/.build/checkouts/swift-nio/Sources/NIOCore/Utilities.swift:45
#4  0x0000aaaaafe750a8 in NIOCore.EventLoopFuture.deinit (self=...) at /MicroSwitch/.build/checkouts/swift-nio/Sources/NIOCore/EventLoopFuture.swift:417
#5  0x0000aaaaafe758dc in NIOCore.EventLoopFuture.__deallocating_deinit (self=...) at /MicroSwitch/.build/checkouts/swift-nio/Sources/NIOCore/EventLoopFuture.swift:416
#6  0x0000ffff9fe8536c in _swift_release_dealloc () from /usr/lib/swift/linux/libswiftCore.so
#7  0x0000ffff9fe85fb0 in bool swift::RefCounts<swift::RefCountBitsT<(swift::RefCountInlinedness)1> >::doDecrementSlow<(swift::PerformDeinit)1>(swift::RefCountBitsT<(swift::RefCountInlinedness)1>, unsigned int) () from /usr/lib/swift/linux/libswiftCore.so
#8  0x0000aaaaafedbbfc in outlined destroy of NIOCore.EventLoopPromise<NIOExtras.(ChannelCollector in _15131B0A1C2B96335CB419BAF83D55DE)> ()
    at /MicroSwitch/.build/checkouts/swift-nio-extras/Sources/NIOExtras/QuiescingHelper.swift:214
#9  0x0000aaaaafedbbd4 in NIOExtras.ServerQuiescingHelper.deinit (self=...) at /MicroSwitch/.build/checkouts/swift-nio-extras/Sources/NIOExtras/QuiescingHelper.swift:214
#10 0x0000aaaaafedbc28 in NIOExtras.ServerQuiescingHelper.__deallocating_deinit (self=...) at /MicroSwitch/.build/checkouts/swift-nio-extras/Sources/NIOExtras/QuiescingHelper.swift:214
#11 0x0000ffff9fe8536c in _swift_release_dealloc () from /usr/lib/swift/linux/libswiftCore.so
#12 0x0000ffff9fe85fb0 in bool swift::RefCounts<swift::RefCountBitsT<(swift::RefCountInlinedness)1> >::doDecrementSlow<(swift::PerformDeinit)1>(swift::RefCountBitsT<(swift::RefCountInlinedness)1>, unsigned int) () from /usr/lib/swift/linux/libswiftCore.so
#13 0x0000aaaaafd4b2a0 in ?? () at /MicroSwitch/.build/checkouts/grpc-swift/Sources/GRPC/Server.swift:199
#14 0x0000ffff9fe8536c in _swift_release_dealloc () from /usr/lib/swift/linux/libswiftCore.so
#15 0x0000ffff9fe85fb0 in bool swift::RefCounts<swift::RefCountBitsT<(swift::RefCountInlinedness)1> >::doDecrementSlow<(swift::PerformDeinit)1>(swift::RefCountBitsT<(swift::RefCountInlinedness)1>, unsigned int) () from /usr/lib/swift/linux/libswiftCore.so
#16 0x0000aaaaafe82364 in ?? () at /MicroSwitch/.build/checkouts/swift-nio/Sources/NIOCore/EventLoopFuture.swift:1570
#17 0x0000ffff9fe8536c in _swift_release_dealloc () from /usr/lib/swift/linux/libswiftCore.so
#18 0x0000ffff9fe85fb0 in bool swift::RefCounts<swift::RefCountBitsT<(swift::RefCountInlinedness)1> >::doDecrementSlow<(swift::PerformDeinit)1>(swift::RefCountBitsT<(swift::RefCountInlinedness)1>, unsigned int) () from /usr/lib/swift/linux/libswiftCore.so
#19 0x0000aaaaafe82228 in ?? () at /MicroSwitch/.build/checkouts/swift-nio/Sources/NIOCore/EventLoopFuture.swift:1570
#20 0x0000ffff9fe8536c in _swift_release_dealloc () from /usr/lib/swift/linux/libswiftCore.so
#21 0x0000ffff9fe85fb0 in bool swift::RefCounts<swift::RefCountBitsT<(swift::RefCountInlinedness)1> >::doDecrementSlow<(swift::PerformDeinit)1>(swift::RefCountBitsT<(swift::RefCountInlinedness)1>, unsigned int) () from /usr/lib/swift/linux/libswiftCore.so
#22 0x0000aaaab00f987c in ?? () at /MicroSwitch/.build/checkouts/swift-nio/Sources/NIOPosix/SelectableEventLoop.swift:418
#23 0x0000ffff9fe8536c in _swift_release_dealloc () from /usr/lib/swift/linux/libswiftCore.so
#24 0x0000ffff9fe85fb0 in bool swift::RefCounts<swift::RefCountBitsT<(swift::RefCountInlinedness)1> >::doDecrementSlow<(swift::PerformDeinit)1>(swift::RefCountBitsT<(swift::RefCountInlinedness)1>, unsigned int) () from /usr/lib/swift/linux/libswiftCore.so
#25 0x0000ffff9fe7aab4 in swift_arrayDestroy () from /usr/lib/swift/linux/libswiftCore.so
#26 0x0000ffff9fc35f58 in Swift.ContiguousArray.replaceSubrange<A where A == A1.Element, A1: Swift.Collection>(_: Swift.Range<Swift.Int>, with: __owned A1) -> () () from /usr/lib/swift/linux/libswiftCore.so
#27 0x0000aaaab00f3ed8 in NIOPosix.SelectableEventLoop.run() throws -> () (self=...) at /MicroSwitch/.build/checkouts/swift-nio/Sources/NIOPosix/SelectableEventLoop.swift:513
#28 0x0000aaaab00c2e0c in static NIOPosix.MultiThreadedEventLoopGroup.(runTheLoop in _C2B1528F4FBA68A3DBFA89DBAEBE9D4D)(thread: NIOPosix.NIOThread, canEventLoopBeShutdownIndividually: Swift.Bool, selectorFactory: () throws -> NIOPosix.Selector<NIOPosix.NIORegistration>, initializer: (NIOPosix.NIOThread) -> (), _: (NIOPosix.SelectableEventLoop) -> ()) -> () (thread=..., canEventLoopBeShutdownIndividually=..., selectorFactory=..., initializer=..., 
    callback=..., self=...) at /MicroSwitch/.build/checkouts/swift-nio/Sources/NIOPosix/MultiThreadedEventLoopGroup.swift:87
#29 0x0000aaaab00c336c in closure #1 (NIOPosix.NIOThread) -> () in static NIOPosix.MultiThreadedEventLoopGroup.(setupThreadAndEventLoop in _C2B1528F4FBA68A3DBFA89DBAEBE9D4D)(name: Swift.String, selectorFactory: () throws -> NIOPosix.Selector<NIOPosix.NIORegistration>, initializer: (NIOPosix.NIOThread) -> ()) -> NIOPosix.SelectableEventLoop () at /MicroSwitch/.build/checkouts/swift-nio/Sources/NIOPosix/MultiThreadedEventLoopGroup.swift:107
#30 0x0000aaaab012a2f8 in reabstraction thunk helper from @escaping @callee_guaranteed (@guaranteed NIOPosix.NIOThread) -> () to @escaping @callee_guaranteed (@in_guaranteed NIOPosix.NIOThread) -> (@out ()) () at /MicroSwitch/.build/checkouts/swift-nio/Sources/NIOPosix/Thread.swift:97
#31 0x0000aaaab012cb84 in closure #1 (Swift.UnsafeMutableRawPointer?) -> Swift.UnsafeMutableRawPointer? in static NIOPosix.ThreadOpsPosix.run(handle: inout Swift.UInt?, args: NIOPosix.Box<(body: (NIOPosix.NIOThread) -> (), name: Swift.String?)>, detachThread: Swift.Bool) -> () ()
    at /MicroSwitch/.build/checkouts/swift-nio/Sources/NIOPosix/ThreadPosix.swift:105
#32 0x0000aaaab012cc94 in @objc closure #1 (Swift.UnsafeMutableRawPointer?) -> Swift.UnsafeMutableRawPointer? in static NIOPosix.ThreadOpsPosix.run(handle: inout Swift.UInt?, args: NIOPosix.Box<(body: (NIOPosix.NIOThread) -> (), name: Swift.String?)>, detachThread: Swift.Bool) -> () ()
    at /MicroSwitch/.build/checkouts/swift-nio/Sources/NIOPosix/ThreadPosix.swift:102
#33 0x0000ffff9f67d5c8 in start_thread (arg=0x0) at ./nptl/pthread_create.c:442
#34 0x0000ffff9f6e5d1c in thread_start () at ../sysdeps/unix/sysv/linux/aarch64/clone.S:79
```

This appears to be being caused by an executed task being dropped ([`#27 0x0000aaaab00f3ed8 in NIOPosix.SelectableEventLoop.run() throws -> () (self=...) at /MicroSwitch/.build/checkouts/swift-nio/Sources/NIOPosix/SelectableEventLoop.swift:513`](https://github.com/apple/swift-nio/blob/cd0aa5907b1e369a129a9134e7ac4c8cac577778/Sources/NIOPosix/SelectableEventLoop.swift#L513)) as a result of event loop shutdown. Is your code shutting down the MTELG?

**Comment by oliverepper** [comment]
Well I have a flag that forces the use of the MTELG on macOS instead of PlatformSupport.makeEventLoopGroup and even with the MTELG it runs fine on macOS. Problem only occurs on Linux. The "listen" function has a defer block that shuts down the group, but only after serving. You can checkout the code: https://github.com/oliverepper/MicroSwitch/blob/main/Sources/MicroSwitch/Server/MicroSwitchServer.swift

**Comment by Lukasa** [comment]
Can you comment out that defer block and confirm your program doesn't immediately exit? Relatedly, can you confirm you're not seeing an error from your logger?

**Comment by oliverepper** [comment]
Same behavior without the defer block.

**Comment by Lukasa** [comment]
(Sorry about the accidental close, mis-clicked.)

That's a surprise to me: I don't see how anyone else could be closing the MTELG! Can you try producing a minimal reproducer from your `main` file? I'll take a quick look at the grpc-swift code and confirm there aren't any secret ELG closes anywhere.

**Comment by Lukasa** [comment]
Yes, grpc-swift doesn't perform any ELG shutdowns on its own. Do you have any other calls to `shutdownGracefully` or `syncShutdownGracefully` in your project that might be being hit?

**Comment by oliverepper** [comment]
No. Only the one in the defer block. I really appreciate your help! If you want you can checkout the complete project, here: https://github.com/oliverepper/MicroSwitch (the first two commands from the README should lead into the error on Linux), I can install MIT license if that helps the matter. I am not expecting support from you it's just that I am a bit helpless at the moment and I know the code did work. So maybe something changed that would be worth a mention, or I am provoking a wired corner case? The project is more or less a demo, it's only about 2k lines of Swift.

**Comment by oliverepper** [comment]
Yes. Actually that was the whole point of the update. To deploy ipv6, finally :-D

**Comment by Lukasa** [comment]
Agreed, I'll repurpose the issue here.

**Comment by Lukasa** [comment]
Oh, no, this issue tracks the dependency bump. I'll create a new issue.

**Comment by Lukasa** [comment]
See #1542.

**Comment by Lukasa** [comment]
Fixed in #1541.
